cruddy/__init__.py

```diff
@@ -52,67 +83,135 @@ def _replace_decimals(self, obj):
+            if encrypted_attr in item:
+                response = self._kms_client.decrypt(
+                    CiphertextBlob=base64.b64decode(item[encrypted_attr]))
+                item[encrypted_attr] = response['Plaintext']
+
     def _check_required(self, item, response):
-        missing = self._required - set(item.keys())
+        missing = set(self.required_attributes) - set(item.keys())
         if len(missing) > 0:
             response['status'] = 'error'
+            response['error_type'] = 'MissingRequiredAttributes'
             response['message'] = 'Missing required attributes: {}'.format(
                 list(missing))
             return False
         return True
 
-    def _list(self, item, response):
-        items = self._table.scan()
-        response['data'] = self._replace_decimals(items['Items'])
-
-    def _create(self, item, response):
-        item['id'] = str(uuid.uuid4())
-        item['created_at'] = int(time.time() * 1000)
-        item['modified_at'] = item['created_at']
-        if self._check_required(item, response):
-            self._table.put_item(Item=item)
-            response['data'] = item
-
-    def _update(self, item, response):
-        item['modified_at'] = int(time.time() * 1000)
-        if self._check_required(item, response):
-            self._table.put_item(Item=item)
-            response['data'] = item
-
-    def _delete(self, item, response):
-        id = item.get('id')
-        if id is None:
+    def _check_supported_op(self, op_name, response):
+        if op_name not in self.supported_ops:
             response['status'] = 'error'
-            response['message'] = 'delete requires an id'
-        else:
-            response['data'] = self._table.delete_item(Key={'id': id})
+            response['error_type'] = 'UnsupportedOperation'
+            response['message'] = 'unsupported operation: {}'.format(op_name)
+            return False
+        return True
 
-    def _get(self, item, response):
-        id = item.get('id')
-        if id is None:
+    def _call_ddb_method(self, method, kwargs, response):
+        try:
+            response['raw_response'] = method(**kwargs)
+        except Exception as e:
             response['status'] = 'error'
-            response['message'] = 'get requires an id'
-        else:
-            item = self._table.get_item(Key={'id': id})['Item']
-            response['data'] = self._replace_decimals(item)
+            response['error_type'] = e.__class__.__name__
+            response['error_message'] = str(e)
+
+    def _new_response(self):
+        return dict(status='success', data=None)
+
+    def _prepare_response(self, response):
+        if response['status'] == 'success':
+            if 'raw_response' in response:
+                md = response['raw_response']['ResponseMetadata']
+                response['response_metadata'] = md
+                del response['raw_response']
```

Comment:
Currently, ``raw_response`` is there if a call was made to DynamoDB and there was an error in the call.  Otherwise, it is stripped out.  I did have a ``debug`` flag that forced it to always be left in there.  I wasn't sure that was useful but I can add it back.  Either way, ``raw_response`` should be documented.